evervault/crypto/client.py

```diff
@@ -37,87 +39,135 @@ def __init__(self, api_key=None, curve="SECP256K1", max_file_size_in_mb=25):
         if type(data) == list:
             encrypted_list = []
             for idx, item in enumerate(data):
                 if not self.__encryptable_data(item):
-                    encrypted_list.insert(idx, self.__traverse_and_encrypt(item))
+                    encrypted_list.insert(
+                        idx, self.__traverse_and_encrypt(item, metadata)
+                    )
                 else:
-                    encrypted_list.insert(idx, self.__encrypt_string(item))
+                    encrypted_list.insert(idx, self.__encrypt_string(item, metadata))
             return encrypted_list
         elif type(data) == dict:
-            return self.__encrypt_object(data)
+            return self.__encrypt_object(data, metadata)
         elif type(data) == set:
-            return self.__encrypt_set(data)
+            return self.__encrypt_set(data, metadata)
         elif self.__encryptable_data(data):
-            return self.__encrypt_string(data)
+            return self.__encrypt_string(data, metadata)
         else:
             raise EvervaultError(f"Cannot encrypt unsupported type {data}")
 
-    def __encrypt_object(self, data):
+    def __encrypt_object(self, data, metadata):
         encrypted_data = {}
         for key, value in data.items():
             if self.__encryptable_data(value):
-                encrypted_data[key] = self.__encrypt_string(value)
+                encrypted_data[key] = self.__encrypt_string(value, metadata)
             else:
-                encrypted_data[key] = self.__traverse_and_encrypt(value)
+                encrypted_data[key] = self.__traverse_and_encrypt(value, metadata)
         return encrypted_data
 
-    def __encrypt_set(self, data):
+    def __encrypt_set(self, data, metadata):
         encrypted_set = set()
         for item in data:
             if self.__encryptable_data(item):
-                encrypted_set.add(self.__encrypt_string(item))
+                encrypted_set.add(self.__encrypt_string(item, metadata))
             else:
-                encrypted_set.add(self.__traverse_and_encrypt(item))
+                encrypted_set.add(self.__traverse_and_encrypt(item, metadata))
         return encrypted_set
 
-    def __encrypt_string(self, data):
+    def __encrypt_string(self, data, metadata):
         header_type = map_header_type(data)
         coerced_data = self.__coerce_type(data)
         iv = token_bytes(12)
         aesgcm = AESGCM(self.shared_key)
+        has_role = "role" in metadata and metadata["role"] is not None
+
+        if has_role:
+            metadata = self.__generate_metadata(metadata["role"])
+            metadata_offset = struct.pack(
+                "<H", len(metadata)
+            )  # '<H' specifies 16-bit unsigned little-endian
+            payload = metadata_offset + metadata + bytes(coerced_data, "utf8")
+        else:
+            payload = bytes(coerced_data, "utf8")
 
         encrypted_bytes = b""
         if self.curve == SECP256K1:
```

Comment:
This is causing the K1 failure for metadata:
```suggestion
        if self.curve == SECP256K1 and not has_role:
```

Reply:
Legend 👑